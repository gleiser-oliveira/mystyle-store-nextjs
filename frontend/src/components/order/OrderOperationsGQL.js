import gql from 'graphql-tag';
import { perPage } from '../../config';

export const SINGLE_ORDER_QUERY = gql`
    query SINGLE_ORDER_QUERY($id: ID!) {
        order(id: $id) {
            id
            charge
            total
            createdAt
            user {
                id
            }
            products {
                id
                title
                description
                price
                image
                quantity
            }
        }
    }
`;

export const USER_ORDERS_QUERY = gql`
    query USER_ORDERS_QUERY($skip: Int = 0, $first: Int = ${perPage}) { 
        orders(first: $first, skip: $skip, orderBy: createdAt_DESC) {
            id
            total
            createdAt
            products {
                id
                title
                price
                description
                quantity
                image
            }
        }
    }
`;

export const USER_ORDERS_PAGINATION_QUERY = gql`
    query USER_ORDERS_PAGINATION_QUERY($userId: ID!) {
        ordersConnection(where: {user: {id: $userId}}) {
            aggregate {
                count
            }
        }
    }
`;
