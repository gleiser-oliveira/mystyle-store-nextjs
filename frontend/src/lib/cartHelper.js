export const numberOfProducts = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.quantity), 0);
}

export const calcTotalCartPrice = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.product.price * b.quantity), 0);
}

export const formatPrice = (priceInCents) => {
    const options = {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
    };
    // if its a whole, dollar amount, leave off the .00
    if (priceInCents % 100 === 0)
        options.minimumFractionDigits = 0;

    const formatter = new Intl.NumberFormat('en-US', options);
    return formatter.format(priceInCents / 100);
}