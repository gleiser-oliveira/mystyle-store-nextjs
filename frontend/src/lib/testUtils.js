import casual from 'casual';
import { act } from 'react-dom/test-utils';
import { CURRENT_USER_QUERY } from '../components/authentication/AuthenticationOperationsGQL';

// seeded for constant results
casual.seed(123);

const fakeUser = overrides => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    email: casual.email,
    permissions: ['USER', 'PRODUCT_CREATE'],
    cart: [],
    ...overrides,
});

const fakeUser2 = () => ({
    __typename: 'User',
    id: '2222',
    name: casual.name,
    email: casual.email,
    permissions: ['USER'],
    orders: [],
    cart: [],
});

const fakeAdminUser = () => ({
    __typename: 'User',
    id: '1111',
    name: casual.name,
    email: casual.email,
    permissions: ['ADMIN', 'PRODUCT_CREATE'],
    orders: [],
    cart: [],
});

const fakeProduct = () => ({
    __typename: 'Product',
    id: 'QWERTY0123',
    title: 'A Fake Product',
    price: 5000,
    description: 'This is a fake product for testing.',
    image: 'product.jpg',
    largeImage: 'largeProduct.jpg',
    user: fakeUser(),
});

const fakeProduct2 = () => ({
    __typename: 'Product',
    id: 'AS012',
    title: 'Another Fake Product',
    price: 2999,
    description: 'This is another fake product for testing.',
    image: 'product2.jpg',
    largeImage: 'bigProduct2.jpg',
    user: fakeUser(),
});

const fakeOrderProduct = () => ({
    __typename: 'OrderProduct',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
});

const fakeOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    products: [fakeOrderProduct(), fakeOrderProduct()],
    createdAt: 'May 1, 2020 00:00:00',
    user: fakeUser(),
});


const fakeCartProduct = overrides => ({
    __typename: 'CartProduct',
    id: 'prod123',
    quantity: 2,
    product: fakeProduct(),
    user: fakeUser(),
    ...overrides,
});

const fakeCartProduct2 = overrides => ({
    __typename: 'CartProduct',
    id: 'prod312',
    quantity: 2,
    product: fakeProduct2(),
    user: fakeUser(),
    ...overrides,
});

const cartUser = fakeUser();
const fakeCart = () => ([
    fakeCartProduct({user: cartUser}),
    fakeCartProduct2({user: cartUser}),
]);

// Fake LocalStorage
class LocalStorageMock {
    constructor() {
        this.store = {};
    }

    clear() {
        this.store = {};
    }

    getItem(key) {
        return this.store[key] || null;
    }

    setItem(key, value) {
        this.store[key] = value.toString();
    }

    removeItem(key) {
        delete this.store[key];
    }
}

// mocks
const signedInMocks = (user = fakeUser) => { return [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: user(),
            }
        }
    }
]};

const notSignedInMocks = () => ([
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: null,
            }
        }
    }
]);

const updateWrapper = async (wrapper, ms = 0) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, ms));
        wrapper.update();
    });
}

export {
    LocalStorageMock,
    fakeProduct,
    fakeProduct2,
    fakeUser,
    fakeUser2,
    fakeAdminUser,
    fakeCartProduct,
    fakeCartProduct2,
    fakeOrder,
    fakeCart,
    fakeOrderProduct,
    signedInMocks,
    notSignedInMocks,
    updateWrapper,
};
