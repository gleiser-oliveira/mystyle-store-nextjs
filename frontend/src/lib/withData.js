import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from './ClientGQL/LocalGQL';
import { DOCKER_BACKEND_URL, BROWSER_TO_BACKEND_URL } from '../config';

function createClient({ headers }) {
    // if using docker, the URI to the Yoga server must be either:
    // DOCKER_BACKEND_URL = the name of the backend server set in the docker network (used in SSR fetches)
    // BROWSER_TO_BACKEND_URL = the URL of the backend server that can be accessed from the client (used in the client's browser fetches)
    const isInsideFrontendServer = typeof window === 'undefined';
    const uri = isInsideFrontendServer ? DOCKER_BACKEND_URL : BROWSER_TO_BACKEND_URL;
    return new ApolloClient({
        uri: uri,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include',
                },
                headers,
            });
        },
        // local data
        clientState: {
            resolvers: {
                Mutation: {
                    toggleCart(_, variables, client) {
                        const { cartOpen } = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        // toggle the state
                        const data = {
                            data: { cartOpen: !cartOpen },
                        };
                        client.cache.writeData(data);
                        return data;
                    },
                },
                Query: {
                    localState(_, variables, client) {
                        const localState = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        return localState;
                    }
                }
            },
            defaults: {
                cartOpen: false,
            }
        }
    });
}

// TODO: check getDataFromTree warning - apollo
export default withApollo(createClient);