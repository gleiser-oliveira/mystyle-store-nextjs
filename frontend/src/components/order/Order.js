import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Loader from 'react-loader-spinner';
import { Query } from 'react-apollo';
import { SINGLE_ORDER_QUERY } from './OrderOperationsGQL';
import { format } from 'date-fns';
import Head from 'next/head';
import { formatPrice } from '../../lib/cartHelper';
import { ErrorMessage } from '../message/Message';
import { Center } from '../theme/generalStyles';
import { OrderStyle } from './styles';

export default class Order extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
    };

    render() {
        return (
            <Query query={SINGLE_ORDER_QUERY} variables={{ id: this.props.id }}>
                {payload => {
                    if (payload.error)
                        return <ErrorMessage error={payload.error} />;
                    if (payload.loading)
                        return (
                            <Center>
                                <Loader type="TailSpin" visible={true} />
                            </Center>
                        );
                    const order = payload.data.order;
                    return (
                        <OrderStyle data-test='order'>
                            <Head>
                                <title>Sick Fits - Order {order.id}</title>
                            </Head>
                            <p>
                                <span>Order ID:</span>
                                <span>{this.props.id}</span>
                            </p>
                            <p>
                                <span>Charge</span>
                                <span>{order.charge}</span>
                            </p>
                            <p>
                                <span>Date</span>
                                <span>{format(new Date(order.createdAt), 'MMMM d, yyyy hh:mm:ss')}</span>
                            </p>
                            <p>
                                <span>Order Total</span>
                                <span>{formatPrice(order.total)}</span>
                            </p>
                            <p>
                                <span>Item Count</span>
                                <span>{order.products.length}</span>
                            </p>
                            <div className="items">
                                {order.products.map(product => (
                                    <div className="order-item" key={product.id}>
                                        <img src={product.image} alt={product.title} />
                                        <div className="item-details">
                                            <h2>{product.title}</h2>
                                            <p>Qty: {product.quantity}</p>
                                            <p>Each: {formatPrice(product.price)}</p>
                                            <p>SubTotal: {formatPrice(product.price * product.quantity)}</p>
                                            <p>{product.description}</p>
                                        </div>
                                    </div>
                                ))}
                            </div>
                        </OrderStyle>
                    );
                }}
            </Query>
        );
    }
}
