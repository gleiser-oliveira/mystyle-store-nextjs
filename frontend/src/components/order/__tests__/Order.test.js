import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import toJSON from 'enzyme-to-json';
import Order from '../Order';
import { SINGLE_ORDER_QUERY } from '../OrderOperationsGQL';
import { updateWrapper, fakeOrder } from '../../../lib/testUtils';

const order = fakeOrder();
const mocks = [
    {
        request: {
            query: SINGLE_ORDER_QUERY,
            variables: { id: order.id }
        },
        result: {
            data: {
                order: {
                    ...order,
                },
            },
        },
    },
];

describe('testing the <RemoveCartProductButton/> component', () => {
    it('renders', async() => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <Order id={order.id}/>
            </MockedProvider>
        );
        await updateWrapper(wrapper);
        const div = wrapper.find('[data-test="order"]');
        expect(toJSON(div)).toMatchSnapshot();
    });
});
