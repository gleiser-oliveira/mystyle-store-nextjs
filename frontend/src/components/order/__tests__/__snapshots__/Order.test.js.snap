// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the <RemoveCartProductButton/> component renders 1`] = `
Array [
  <styles__OrderStyle
    data-test="order"
  >
    <div
      className="styles__OrderStyle-yugidx-0 cCeASZ"
      data-test="order"
    >
      <Head>
        <Component
          handleStateChange={null}
          inAmpMode={false}
          reduceComponentsToState={[Function]}
        />
      </Head>
      <p>
        <span>
          Order ID:
        </span>
        <span>
          ord123
        </span>
      </p>
      <p>
        <span>
          Charge
        </span>
        <span>
          ch_123
        </span>
      </p>
      <p>
        <span>
          Date
        </span>
        <span>
          May 1, 2020 12:00:00
        </span>
      </p>
      <p>
        <span>
          Order Total
        </span>
        <span>
          $400
        </span>
      </p>
      <p>
        <span>
          Item Count
        </span>
        <span>
          2
        </span>
      </p>
      <div
        className="items"
      >
        <div
          className="order-item"
          key="7cf6a972-765a-4b57-80b6-8ba27218b86a"
        >
          <img
            alt="pariatur quos est quod laborum ea quisquam"
            src="ullam.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
              pariatur quos est quod laborum ea quisquam
            </h2>
            <p>
              Qty: 
              1
            </p>
            <p>
              Each: 
              $42.34
            </p>
            <p>
              SubTotal: 
              $42.34
            </p>
            <p>
              esse quia et commodi autem incidunt ut
            </p>
          </div>
        </div>
        <div
          className="order-item"
          key="5fac1c60-7272-4614-b6ce-af8879914576"
        >
          <img
            alt="et repudiandae magnam aut atque dolores voluptatibus"
            src="soluta.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
              et repudiandae magnam aut atque dolores voluptatibus
            </h2>
            <p>
              Qty: 
              1
            </p>
            <p>
              Each: 
              $42.34
            </p>
            <p>
              SubTotal: 
              $42.34
            </p>
            <p>
              ut occaecati iusto laborum placeat quia deleniti
            </p>
          </div>
        </div>
      </div>
    </div>
  </styles__OrderStyle>,
  <div
    className="styles__OrderStyle-yugidx-0 cCeASZ"
    data-test="order"
  >
    <Head>
      <Component
        handleStateChange={null}
        inAmpMode={false}
        reduceComponentsToState={[Function]}
      />
    </Head>
    <p>
      <span>
        Order ID:
      </span>
      <span>
        ord123
      </span>
    </p>
    <p>
      <span>
        Charge
      </span>
      <span>
        ch_123
      </span>
    </p>
    <p>
      <span>
        Date
      </span>
      <span>
        May 1, 2020 12:00:00
      </span>
    </p>
    <p>
      <span>
        Order Total
      </span>
      <span>
        $400
      </span>
    </p>
    <p>
      <span>
        Item Count
      </span>
      <span>
        2
      </span>
    </p>
    <div
      className="items"
    >
      <div
        className="order-item"
        key="7cf6a972-765a-4b57-80b6-8ba27218b86a"
      >
        <img
          alt="pariatur quos est quod laborum ea quisquam"
          src="ullam.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            pariatur quos est quod laborum ea quisquam
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            SubTotal: 
            $42.34
          </p>
          <p>
            esse quia et commodi autem incidunt ut
          </p>
        </div>
      </div>
      <div
        className="order-item"
        key="5fac1c60-7272-4614-b6ce-af8879914576"
      >
        <img
          alt="et repudiandae magnam aut atque dolores voluptatibus"
          src="soluta.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            et repudiandae magnam aut atque dolores voluptatibus
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            SubTotal: 
            $42.34
          </p>
          <p>
            ut occaecati iusto laborum placeat quia deleniti
          </p>
        </div>
      </div>
    </div>
  </div>,
]
`;
